src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp

```diff
@@ -265,59 +369,32 @@ void ov::npuw::LLMInferRequest::infer_generate(ov::SoPtr<ov::ITensor> input_ids,
-    if (kvcache_desc.num_stored_tokens == kvcache_desc.total_size) {
-        OPENVINO_THROW("KV-Cache is full.");
-    }
-
-    if (m_need_copy_kvcache) {
-        LOG_DEBUG("Copying kv-cache from prefill to generate model.");
-        const std::size_t kStartOutputKVCacheLayers = 1u;
-        const auto& kvcache_compiled = m_kvcache_request->get_compiled_model();
-        // FIXME: Find only matching by names outputs and copy them, having previously checked that such inputs exist
-        for (std::size_t i = 0; i < kvcache_compiled->outputs().size() - 1; ++i) {
-            const auto& output_name = kvcache_compiled->outputs()[kStartOutputKVCacheLayers + i].get_any_name();
-            auto prefill_out_tensor = m_prefill_request->get_tensor(m_prefill_out_ports.at(output_name));
-
-            const auto& input_name = std::regex_replace(output_name, std::regex("present"), "past_key_values");
-            if (m_kvcache_in_ports.find(input_name) == m_kvcache_in_ports.end()) {
-                LOG_DEBUG("Input name " << input_name << " doesn't contain kv cache. Skipping.");
-                continue;
-            }
-
-            auto kvcache_in_tensor = m_kvcache_request->get_tensor(m_kvcache_in_ports.at(input_name));
-
-            // FIXME: We don't need to fill whole tensor with 0s, but only tensor.size() - num_stored_tokens
-            //        taking into account kvcache dimension.
-            fill_tensor<ov::float16>(kvcache_in_tensor, 0);
+    // TODO: Use another solution than flag if multithreading comes to play.
+    if (!m_generate_initialized) {
+        LOG_DEBUG("Copy kv-cache asynchronously...");
+        auto copy_kvcache_future = std::async(std::launch::async, &LLMInferRequest::copy_kvcache, this);
 
-            const auto& kv_dim = (output_name.find("value") != std::string::npos && kvcache_desc.v_tensors_transposed)
-                                     ? 3u
-                                     : kvcache_desc.dim;
-
-            auto prefill_out_slice = make_tensor_slice(prefill_out_tensor,
-                                                       kv_dim,
-                                                       kvcache_desc.max_prompt_size - kvcache_desc.num_stored_tokens,
-                                                       kvcache_desc.max_prompt_size);
-
-            auto kvcache_in_slice = make_tensor_slice(kvcache_in_tensor, kv_dim, 0u, kvcache_desc.num_stored_tokens);
-
-            if (kv_dim == 3u) {
-                copy_columns_by_row_chunks(prefill_out_slice, kvcache_in_slice);
-            } else if (kv_dim == 2u) {
-                copy_by_planes(prefill_out_slice, kvcache_in_slice);
-            } else {
-                prefill_out_slice->copy_to(kvcache_in_slice._ptr);
-            }
+        LOG_DEBUG("Prepare attention mask pattern.");
+        auto kv_attn_mask = m_kvcache_request->get_tensor(m_kvcache_in_ports.at("attention_mask"));
+        fill_tensor<int64_t>(kv_attn_mask, 0);
```

Comment:
why don't use a plain memset? Is `fill_tensor` more complex than that? I believe attention_mask isn't a strided view tensor on top of something, isn't it?

Reply:
It shouldn't be per my understanding... This code was just ported from GenAI, I think